# In this project, I worked with a dataset of used cars from eBay Kleinanzeigen, a classified section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. Dataquest made a few modifications from the original dataset that was uploaded to Kaggle:

They sampled 50,000 data points from the full dataset, to ensure the code runs quickly on local machine.


The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## Cleaning column names

- Reordering the columns
- Convert the column names from camelcase to snakecase 
- Reword some column names based on the data dictionary to be more descriptive.

In [6]:
autos = autos[[
        'dateCrawled', 'dateCreated', 'lastSeen', 'yearOfRegistration', 'monthOfRegistration', 'name', 
        'abtest', 'vehicleType', 'gearbox', 'powerPS', 'brand', 'model', 'odometer', 'fuelType',
        'notRepairedDamage', 'postalCode', 'seller', 'offerType', 'nrOfPictures', 'price'
       ]]

In [7]:
autos.columns = [
        'date_crawled', 'date_created', 'last_seen', 'registration_year', 'registration_month', 'name', 
        'abtest', 'vehicle_type', 'gearbox', 'power_ps', 'brand', 'model', 'odometer', 'fuel_type',
        'unrepaired_damage', 'postal_code', 'seller', 'offer_type', 'nr_of_pictures', 'price'
       ]

In [8]:
autos.head()

date_crawled         date_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  registration_month  \
0               2004                   3   
1               1997                   6   
2               2009                   7   
3               2007                   6   
4               2003                   7   

                                                name   abtest vehicle_type  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  control          bus   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  control    limousine   
2                         Volkswagen_Golf_1.6_United     test    limousine   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  control   kleinwagen   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...     test        kombi   

     gearbox  power_ps       brand   model   odometer fuel_type  \
0    manuell       158     peugeot  andere  150,000km       lpg   
1  automatik       286         bmw     7er  150,000km    benzin   
2    manuell       102  volkswagen    golf   70,000km    benzin   
3  automatik        71       smart  fortwo   70,000km    benzin   
4    manuell         0        ford   focus  150,000km    benzin   

  unrepaired_damage  postal_code  seller offer_type  nr_of_pictures   price  
0              nein        79588  privat    Angebot               0  $5,000  
1              nein        71034  privat    Angebot               0  $8,500  
2              nein        35394  privat    Angebot               0  $8,990  
3              nein        33729  privat    Angebot               0  $4,350  
4              nein        39218  privat    Angebot               0  $1,350

## Initial exploration and cleaning

Let's look at descriptive statistics for all columns.

In [9]:
autos.describe(include='all')

date_crawled         date_created            last_seen  \
count                 50000                50000                50000   
unique                48213                   76                39481   
top     2016-03-21 20:37:19  2016-04-03 00:00:00  2016-04-07 06:17:27   
freq                      3                 1946                    8   
mean                    NaN                  NaN                  NaN   
std                     NaN                  NaN                  NaN   
min                     NaN                  NaN                  NaN   
25%                     NaN                  NaN                  NaN   
50%                     NaN                  NaN                  NaN   
75%                     NaN                  NaN                  NaN   
max                     NaN                  NaN                  NaN   

        registration_year  registration_month         name abtest  \
count        50000.000000        50000.000000        50000  50000   
unique                NaN                 NaN        38754      2   
top                   NaN                 NaN  Ford_Fiesta   test   
freq                  NaN                 NaN           78  25756   
mean          2005.073280            5.723360          NaN    NaN   
std            105.712813            3.711984          NaN    NaN   
min           1000.000000            0.000000          NaN    NaN   
25%           1999.000000            3.000000          NaN    NaN   
50%           2003.000000            6.000000          NaN    NaN   
75%           2008.000000            9.000000          NaN    NaN   
max           9999.000000           12.000000          NaN    NaN   

       vehicle_type  gearbox      power_ps       brand  model   odometer  \
count         44905    47320  50000.000000       50000  47242      50000   
unique            8        2           NaN          40    245         13   
top       limousine  manuell           NaN  volkswagen   golf  150,000km   
freq          12859    36993           NaN       10687   4024      32424   
mean            NaN      NaN    116.355920         NaN    NaN        NaN   
std             NaN      NaN    209.216627         NaN    NaN        NaN   
min             NaN      NaN      0.000000         NaN    NaN        NaN   
25%             NaN      NaN     70.000000         NaN    NaN        NaN   
50%             NaN      NaN    105.000000         NaN    NaN        NaN   
75%             NaN      NaN    150.000000         NaN    NaN        NaN   
max             NaN      NaN  17700.000000         NaN    NaN        NaN   

       fuel_type unrepaired_damage   postal_code  seller offer_type  \
count      45518             40171  50000.000000   50000      50000   
unique         7                 2           NaN       2          2   
top       benzin              nein           NaN  privat    Angebot   
freq       30107             35232           NaN   49999      49999   
mean         NaN               NaN  50813.627300     NaN        NaN   
std          NaN               NaN  25779.747957     NaN        NaN   
min          NaN               NaN   1067.000000     NaN        NaN   
25%          NaN               NaN  30451.000000     NaN        NaN   
50%          NaN               NaN  49577.000000     NaN        NaN   
75%          NaN               NaN  71540.000000     NaN        NaN   
max          NaN               NaN  99998.000000     NaN        NaN   

        nr_of_pictures  price  
count          50000.0  50000  
unique             NaN   2357  
top                NaN     $0  
freq               NaN   1421  
mean               0.0    NaN  
std                0.0    NaN  
min                0.0    NaN  
25%                0.0    NaN  
50%                0.0    NaN  
75%                0.0    NaN  
max                0.0    NaN

### Columns that needs to be altered

- seller and offer_type have mostly one value with count 49999
- nr_of_pictures has only 0 value
- price and odometer columns are numeric values stored as text.

Remove the "seller", "offer_type", and "nr_of_pictures" columns

In [10]:
autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis = 1, inplace = True)

In [11]:
autos.head()

date_crawled         date_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  registration_month  \
0               2004                   3   
1               1997                   6   
2               2009                   7   
3               2007                   6   
4               2003                   7   

                                                name   abtest vehicle_type  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  control          bus   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  control    limousine   
2                         Volkswagen_Golf_1.6_United     test    limousine   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  control   kleinwagen   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...     test        kombi   

     gearbox  power_ps       brand   model   odometer fuel_type  \
0    manuell       158     peugeot  andere  150,000km       lpg   
1  automatik       286         bmw     7er  150,000km    benzin   
2    manuell       102  volkswagen    golf   70,000km    benzin   
3  automatik        71       smart  fortwo   70,000km    benzin   
4    manuell         0        ford   focus  150,000km    benzin   

  unrepaired_damage  postal_code   price  
0              nein        79588  $5,000  
1              nein        71034  $8,500  
2              nein        35394  $8,990  
3              nein        33729  $4,350  
4              nein        39218  $1,350

Convert "price" and "odometer" columns into numeric columns

In [12]:
autos['price'] = autos['price'].str.replace(",","").str.replace("$","")
autos['price'] = autos['price'].astype('int')

autos['odometer'] = autos['odometer'].str.replace("km","").str.replace(",","")
autos['odometer'] = autos['odometer'].astype('int')


In [13]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace = True)

In [14]:
autos[['price', 'odometer_km']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
price          50000 non-null int64
odometer_km    50000 non-null int64
dtypes: int64(2)
memory usage: 781.4 KB


Some data are stored in German words, let's convert them into English.

In [15]:
autos.head()

date_crawled         date_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  registration_month  \
0               2004                   3   
1               1997                   6   
2               2009                   7   
3               2007                   6   
4               2003                   7   

                                                name   abtest vehicle_type  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  control          bus   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  control    limousine   
2                         Volkswagen_Golf_1.6_United     test    limousine   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  control   kleinwagen   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...     test        kombi   

     gearbox  power_ps       brand   model  odometer_km fuel_type  \
0    manuell       158     peugeot  andere       150000       lpg   
1  automatik       286         bmw     7er       150000    benzin   
2    manuell       102  volkswagen    golf        70000    benzin   
3  automatik        71       smart  fortwo        70000    benzin   
4    manuell         0        ford   focus       150000    benzin   

  unrepaired_damage  postal_code  price  
0              nein        79588   5000  
1              nein        71034   8500  
2              nein        35394   8990  
3              nein        33729   4350  
4              nein        39218   1350

In [16]:
autos['abtest'].unique()

array(['control', 'test'], dtype=object)

In [17]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [18]:
autos['vehicle_type'] = autos['vehicle_type'].str.replace("kombi","wagon")
autos['vehicle_type'] = autos['vehicle_type'].str.replace("kleinwagen","compact")
autos['vehicle_type'] = autos['vehicle_type'].str.replace("cabrio","convertible")
autos['vehicle_type'] = autos['vehicle_type'].str.replace("andere","other")
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'compact', 'wagon', nan, 'coupe', 'suv',
       'convertible', 'other'], dtype=object)

In [19]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [20]:
autos['gearbox'] = autos['gearbox'].str.replace("manuell","manual")
autos['gearbox'] = autos['gearbox'].str.replace("automatik","automatic")
autos['gearbox'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [21]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [22]:
autos['fuel_type'] = autos['fuel_type'].str.replace("benzin","gasoline")
autos['fuel_type'] = autos['fuel_type'].str.replace("andere","other")
autos['fuel_type'] = autos['fuel_type'].str.replace("elektro","electro")
autos['fuel_type'].unique()

array(['lpg', 'gasoline', 'diesel', nan, 'cng', 'hybrid', 'electro',
       'other'], dtype=object)

In [23]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [24]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("nein","no")
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace("ja","yes")
autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)

Convert the dates to be uniform numeric data

In [25]:
autos['date_crawled'] = autos['date_crawled'].str.replace("-","")
autos['date_created'] = autos['date_created'].str.replace("-","")
autos['last_seen'] = autos['last_seen'].str.replace("-","")


In [26]:
autos.head()

date_crawled       date_created          last_seen  registration_year  \
0  20160326 17:47:46  20160326 00:00:00  20160406 06:45:54               2004   
1  20160404 13:38:56  20160404 00:00:00  20160406 14:45:08               1997   
2  20160326 18:57:24  20160326 00:00:00  20160406 20:15:37               2009   
3  20160312 16:58:10  20160312 00:00:00  20160315 03:16:28               2007   
4  20160401 14:38:50  20160401 00:00:00  20160401 14:38:50               2003   

   registration_month                                               name  \
0                   3                   Peugeot_807_160_NAVTECH_ON_BOARD   
1                   6         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2                   7                         Volkswagen_Golf_1.6_United   
3                   6  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4                   7  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    abtest vehicle_type    gearbox  power_ps       brand   model  odometer_km  \
0  control          bus     manual       158     peugeot  andere       150000   
1  control    limousine  automatic       286         bmw     7er       150000   
2     test    limousine     manual       102  volkswagen    golf        70000   
3  control      compact  automatic        71       smart  fortwo        70000   
4     test        wagon     manual         0        ford   focus       150000   

  fuel_type unrepaired_damage  postal_code  price  
0       lpg                no        79588   5000  
1  gasoline                no        71034   8500  
2  gasoline                no        35394   8990  
3  gasoline                no        33729   4350  
4  gasoline                no        39218   1350

## Exploring the Odometer and Price Columns

In [27]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [28]:
autos['price'].value_counts().sort_index(ascending = False).head(9)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
Name: price, dtype: int64

Find and remove outliers in "price" column using Z-Score

In [29]:
z = np.abs(stats.zscore(autos['price']))
price_outside_threshold = np.where(z > 3)[0]
price_outside_threshold

array([ 2897, 11137, 24384, 27371, 39377, 39705, 42221, 47598, 47634])

In [30]:
autos.drop(price_outside_threshold, inplace = True)

In [31]:
autos['price'].describe()

count    4.999100e+04
mean     5.831664e+03
std      1.427337e+04
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.300000e+06
Name: price, dtype: float64

Let's take a look at "odometer_km" column

In [32]:
autos['odometer_km'].value_counts()

150000    32419
125000     5170
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1026
5000        966
40000       818
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [33]:
autos['odometer_km'].describe()

count     49991.000000
mean     125736.432558
std       40038.005358
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [34]:
z = np.abs(stats.zscore(autos['odometer_km']))
odometer_outside_threshold = np.where(z > 3)[0]

In [35]:
new_df = autos.drop(odometer_outside_threshold)
new_df['odometer_km'].describe()

count     49025.000000
mean     125885.262621
std       39818.242098
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

After removing the outliers from "odometer_km", it doesn't make a difference from a statistical perspective. Therefore I chose to keep them.


## Exploring the date columns

In [36]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

20160305 1    0.018723
20160305 2    0.006661
20160306 0    0.002821
20160306 1    0.008321
20160306 2    0.002801
                ...   
20160406 0    0.000840
20160406 1    0.001740
20160406 2    0.000600
20160407 0    0.000720
20160407 1    0.000700
Name: date_crawled, Length: 100, dtype: float64

In [37]:
autos['date_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

20150611 0    0.000020
20150810 0    0.000020
20150909 0    0.000020
20151110 0    0.000020
20151205 0    0.000020
                ...   
20160403 0    0.038927
20160404 0    0.036867
20160405 0    0.011842
20160406 0    0.003261
20160407 0    0.001280
Name: date_created, Length: 76, dtype: float64

In [38]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

20160305 1    0.000720
20160305 2    0.000360
20160306 0    0.003081
20160306 1    0.001220
20160306 2    0.000120
                ...   
20160406 0    0.095177
20160406 1    0.089416
20160406 2    0.036387
20160407 0    0.093097
20160407 1    0.037847
Name: last_seen, Length: 100, dtype: float64

In [39]:
autos['registration_year'].describe()

count    49991.000000
mean      2005.074533
std        105.721987
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Dealing with incorrect "registration_year" data

Any vehicle with a registration year above 2016 is definitely inaccurate because a car must be registered before its listing. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [40]:
autos = autos[autos['registration_year'].between(1990,2016)]
autos['registration_year'].describe()

count    46631.000000
mean      2003.517810
std          5.801785
min       1990.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

## Exploring Brands

Top 20 brands


In [41]:
top20brands = autos['brand'].value_counts(normalize=True).head(20)
top20brands

volkswagen       0.211748
bmw              0.111235
opel             0.109391
mercedes_benz    0.092664
audi             0.087860
ford             0.069868
renault          0.048423
peugeot          0.030388
fiat             0.025863
seat             0.018700
skoda            0.016405
mazda            0.015483
nissan           0.015355
smart            0.014325
citroen          0.013918
toyota           0.012631
hyundai          0.010143
volvo            0.009028
mini             0.008792
mitsubishi       0.008342
Name: brand, dtype: float64

### Finding the mean price of each brand

In [42]:
brand_mean_prices = {}

for brand in top20brands.index:    
    priceLogic = autos['brand'] == brand
    priceMean = autos[priceLogic]['price'].mean()
    brand_mean_prices[brand] = priceMean
brand_mean_prices 

{'volkswagen': 5233.7299979744785,
 'bmw': 8372.065741276268,
 'opel': 2859.1368359145267,
 'mercedes_benz': 8447.277944920157,
 'audi': 9181.35147669026,
 'ford': 3668.304481276857,
 'renault': 2342.2395925597875,
 'peugeot': 3033.850388143966,
 'fiat': 2610.973466003317,
 'seat': 4301.417431192661,
 'skoda': 6360.520261437909,
 'mazda': 4028.02216066482,
 'nissan': 4687.019553072626,
 'smart': 3542.706586826347,
 'citroen': 3597.144838212635,
 'toyota': 5114.405772495756,
 'hyundai': 5308.53911205074,
 'volvo': 4698.788598574822,
 'mini': 10534.160975609757,
 'mitsubishi': 3338.60411311054}

In [43]:
bmp_series = pd.Series(brand_mean_prices)
bmp_series.head()

volkswagen       5233.729998
bmw              8372.065741
opel             2859.136836
mercedes_benz    8447.277945
audi             9181.351477
dtype: float64

In [44]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df.head()

mean_price
volkswagen     5233.729998
bmw            8372.065741
opel           2859.136836
mercedes_benz  8447.277945
audi           9181.351477

### Finding the mean mileage of each brand

In [45]:
brand_mean_mileage = {}

for brand in top20brands.index:    
    mileageLogic = autos['brand'] == brand
    mileageMean = autos[mileageLogic]['odometer_km'].mean()
    brand_mean_mileage[brand] = mileageMean
brand_mean_mileage

{'volkswagen': 129127.5065829451,
 'bmw': 132565.06651243495,
 'opel': 129836.30660654773,
 'mercedes_benz': 130925.71164082388,
 'audi': 129090.79814498413,
 'ford': 125164.21117249847,
 'renault': 128461.02745792737,
 'peugeot': 127191.24911785462,
 'fiat': 117682.42122719735,
 'seat': 121697.24770642201,
 'skoda': 111267.97385620915,
 'mazda': 124833.79501385041,
 'nissan': 118386.87150837989,
 'smart': 99595.80838323354,
 'citroen': 120354.39137134052,
 'toyota': 115585.73853989813,
 'hyundai': 106511.62790697675,
 'volvo': 139477.43467933493,
 'mini': 88609.75609756098,
 'mitsubishi': 127069.4087403599}

In [46]:
bmm_series = pd.Series(brand_mean_mileage)
bmm_series.head()

volkswagen       129127.506583
bmw              132565.066512
opel             129836.306607
mercedes_benz    130925.711641
audi             129090.798145
dtype: float64

In [47]:
df['mean_mileage_km'] = bmm_series
df.head()

mean_price  mean_mileage_km
volkswagen     5233.729998    129127.506583
bmw            8372.065741    132565.066512
opel           2859.136836    129836.306607
mercedes_benz  8447.277945    130925.711641
audi           9181.351477    129090.798145

### Finding the most common brand/model combinations


In [48]:
brand = autos['brand']
model = autos['model']
autos['brand_model'] = autos['brand'] + '_' + autos['model']
autos['brand_model'].value_counts()

volkswagen_golf      3707
bmw_3er              2644
volkswagen_polo      1661
opel_corsa           1642
opel_astra           1387
                     ... 
land_rover_andere       1
rover_rangerover        1
ford_b_max              1
rover_discovery         1
bmw_i3                  1
Name: brand_model, Length: 289, dtype: int64

### Spliting Odometer into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

In [49]:
lessthan100km = autos['odometer_km'].between(0,100000)
autos[lessthan100km]['price'].describe()

count    1.141700e+04
mean     1.143786e+04
std      1.867922e+04
min      0.000000e+00
25%      3.400000e+03
50%      7.900000e+03
75%      1.529900e+04
max      1.300000e+06
Name: price, dtype: float64

In [50]:
between101km_and_125km = autos['odometer_km'].between(100001,125000)
autos[between101km_and_125km]['price'].describe()

count      4845.000000
mean       6272.256553
std       15898.336670
min           0.000000
25%        1650.000000
50%        3990.000000
75%        8150.000000
max      999999.000000
Name: price, dtype: float64

In [51]:
morethan125km = autos['odometer_km'].between(125001,150000)
autos[morethan125km]['price'].describe()

count    3.036900e+04
mean     3.650193e+03
std      8.437727e+03
min      0.000000e+00
25%      9.000000e+02
50%      2.000000e+03
75%      4.500000e+03
max      1.234566e+06
Name: price, dtype: float64

As the descriptive data shows above, we can see that more mileages comes with lower prices. 

### How much more expensive are cars with undamaged than their damaged counterparts?

In [52]:
autos.head()

date_crawled       date_created          last_seen  registration_year  \
0  20160326 17:47:46  20160326 00:00:00  20160406 06:45:54               2004   
1  20160404 13:38:56  20160404 00:00:00  20160406 14:45:08               1997   
2  20160326 18:57:24  20160326 00:00:00  20160406 20:15:37               2009   
3  20160312 16:58:10  20160312 00:00:00  20160315 03:16:28               2007   
4  20160401 14:38:50  20160401 00:00:00  20160401 14:38:50               2003   

   registration_month                                               name  \
0                   3                   Peugeot_807_160_NAVTECH_ON_BOARD   
1                   6         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2                   7                         Volkswagen_Golf_1.6_United   
3                   6  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4                   7  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    abtest vehicle_type    gearbox  power_ps       brand   model  odometer_km  \
0  control          bus     manual       158     peugeot  andere       150000   
1  control    limousine  automatic       286         bmw     7er       150000   
2     test    limousine     manual       102  volkswagen    golf        70000   
3  control      compact  automatic        71       smart  fortwo        70000   
4     test        wagon     manual         0        ford   focus       150000   

  fuel_type unrepaired_damage  postal_code  price      brand_model  
0       lpg                no        79588   5000   peugeot_andere  
1  gasoline                no        71034   8500          bmw_7er  
2  gasoline                no        35394   8990  volkswagen_golf  
3  gasoline                no        33729   4350     smart_fortwo  
4  gasoline                no        39218   1350       ford_focus

In [53]:
comparedamage = autos[['unrepaired_damage', 'brand_model', 'price']].dropna()
comparedamage.head()

unrepaired_damage      brand_model  price
0                no   peugeot_andere   5000
1                no          bmw_7er   8500
2                no  volkswagen_golf   8990
3                no     smart_fortwo   4350
4                no       ford_focus   1350

In [54]:
undamagedLogic = autos['unrepaired_damage'] == 'no'
damagedLogic = autos['unrepaired_damage'] == 'yes'

In [55]:
undamaged = comparedamage[undamagedLogic]
undamaged.head()

unrepaired_damage      brand_model  price
0                no   peugeot_andere   5000
1                no          bmw_7er   8500
2                no  volkswagen_golf   8990
3                no     smart_fortwo   4350
4                no       ford_focus   1350

In [56]:
damaged = comparedamage[damagedLogic]
damaged.head()

unrepaired_damage             brand_model  price
28               yes             mini_cooper   5250
30               yes         peugeot_2_reihe     80
51               yes  mercedes_benz_s_klasse   6000
80               yes          nissan_primera      0
81               yes            nissan_micra   2000

Two points that should be noted:
- We're only comparing only if the combination exists in both damaged and undamaged conditions 
- We're comparing the mean price

In [57]:
undamagedic = {}
damagedic = {}
unlist = [i for i in undamaged['brand_model']]
damlist = [i for i in damaged['brand_model']]
for i in unlist:
    if i in damlist:
        undamagedic[i] = undamaged[undamaged['brand_model'] == i]['price'].mean()
        damagedic[i] = damaged[damaged['brand_model'] == i]['price'].mean()

In [58]:
damagedic

{'peugeot_andere': 2079.2,
 'bmw_7er': 5070.846153846154,
 'volkswagen_golf': 1766.5284810126582,
 'smart_fortwo': 1507.1025641025642,
 'ford_focus': 1582.0860215053763,
 'seat_arosa': 543.6875,
 'renault_megane': 1101.7142857142858,
 'audi_a3': 2728.720588235294,
 'opel_vectra': 878.8658536585366,
 'mazda_andere': 1279.2307692307693,
 'porsche_911': 20833.333333333332,
 'mini_cooper': 4271.111111111111,
 'bmw_5er': 3736.7830188679245,
 'volkswagen_polo': 942.968085106383,
 'mercedes_benz_e_klasse': 4425.7692307692305,
 'mercedes_benz_c_klasse': 3739.635135135135,
 'ford_mondeo': 1021.9651162790698,
 'seat_altea': 3133.3333333333335,
 'renault_twingo': 741.1973684210526,
 'toyota_andere': 2410.0,
 'mercedes_benz_a_klasse': 1610.0727272727272,
 'mercedes_benz_cl': 14425.0,
 'peugeot_3_reihe': 1256.6444444444444,
 'dacia_sandero': 4219.25,
 'volkswagen_passat': 2070.271186440678,
 'nissan_primera': 999.6111111111111,
 'opel_corsa': 1135.0344827586207,
 'saab_andere': 574.75,
 'audi_a6': 

In [59]:
undamagedic

{'peugeot_andere': 5281.702127659574,
 'bmw_7er': 10345.53409090909,
 'volkswagen_golf': 6041.277108433735,
 'smart_fortwo': 4011.6495327102803,
 'ford_focus': 5841.520426287744,
 'seat_arosa': 1137.967741935484,
 'renault_megane': 4046.563559322034,
 'audi_a3': 9176.349606299213,
 'opel_vectra': 2194.8672566371683,
 'mazda_andere': 3213.782608695652,
 'porsche_911': 74406.50847457627,
 'mini_cooper': 11967.05138339921,
 'bmw_5er': 8870.08608490566,
 'volkswagen_polo': 3174.635531135531,
 'mercedes_benz_e_klasse': 9304.964234620888,
 'mercedes_benz_c_klasse': 7686.884742041712,
 'ford_mondeo': 3802.6485623003196,
 'seat_altea': 6376.472222222223,
 'renault_twingo': 1726.847290640394,
 'toyota_andere': 6506.443037974684,
 'mercedes_benz_a_klasse': 4536.6560975609755,
 'mercedes_benz_cl': 12802.58064516129,
 'peugeot_3_reihe': 3511.101694915254,
 'dacia_sandero': 5666.333333333333,
 'volkswagen_passat': 5709.010080645161,
 'nissan_primera': 1603.7142857142858,
 'opel_corsa': 2239.3092979

Showing the price difference between undamaged and damaged cars

In [60]:
price_diff = {}
for br_md in damagedic:
    price_diff[br_md] = undamagedic[br_md] - damagedic[br_md]

price_diff

{'peugeot_andere': 3202.5021276595744,
 'bmw_7er': 5274.687937062936,
 'volkswagen_golf': 4274.748627421077,
 'smart_fortwo': 2504.546968607716,
 'ford_focus': 4259.434404782368,
 'seat_arosa': 594.2802419354839,
 'renault_megane': 2944.8492736077483,
 'audi_a3': 6447.629018063919,
 'opel_vectra': 1316.0014029786316,
 'mazda_andere': 1934.5518394648827,
 'porsche_911': 53573.175141242944,
 'mini_cooper': 7695.940272288099,
 'bmw_5er': 5133.303066037736,
 'volkswagen_polo': 2231.667446029148,
 'mercedes_benz_e_klasse': 4879.195003851657,
 'mercedes_benz_c_klasse': 3947.2496069065774,
 'ford_mondeo': 2780.68344602125,
 'seat_altea': 3243.138888888889,
 'renault_twingo': 985.6499222193414,
 'toyota_andere': 4096.443037974684,
 'mercedes_benz_a_klasse': 2926.5833702882483,
 'mercedes_benz_cl': -1622.4193548387102,
 'peugeot_3_reihe': 2254.4572504708094,
 'dacia_sandero': 1447.083333333333,
 'volkswagen_passat': 3638.738894204483,
 'nissan_primera': 604.1031746031747,
 'opel_corsa': 1104.27

In [61]:
diff = 0
for br_md in damagedic:
    diff += undamagedic[br_md] - damagedic[br_md]
    
avg_diff = diff / len(damagedic)

print ("The average price difference between undamaged and damaged car is ${}".format(avg_diff) )

The average price difference between undamaged and damaged car is $3796.459112000242


### Brand_Model that are more expensive in damaged condition than its undamaged counterparts

In [62]:
dmg_exp = []
for br_md in damagedic:
    diff = undamagedic[br_md] - damagedic[br_md]
    if diff < 0:
        dmg_exp.append(br_md)
dmg_exp

['mercedes_benz_cl',
 'skoda_superb',
 'volkswagen_touareg',
 'mercedes_benz_viano',
 'alfa_romeo_159',
 'lancia_lybra',
 'lancia_andere']

### TODO: Further exploring the price difference

There might be some features that affect the price

In [63]:
df = autos[autos['brand_model'] == 'peugeot_andere'].dropna()
columns = ['registration_year', 'gearbox', 'power_ps', 'unrepaired_damage', 'odometer_km', 'price']
df[columns].sort_values(by = 'registration_year').tail(50)

registration_year    gearbox  power_ps unrepaired_damage  odometer_km  \
23401               2006     manual       109                no       150000   
49684               2006     manual       128                no       150000   
34750               2006     manual        58                no        80000   
8982                2006     manual       120                no        60000   
10484               2006     manual        84                no       150000   
39202               2006  automatic       204               yes       150000   
7999                2007     manual       134               yes       150000   
7022                2007     manual       109                no       125000   
48136               2007     manual        95                no       125000   
33752               2007     manual         0                no       125000   
46317               2008     manual         0               yes        60000   
22176               2008     manual       120                no       100000   
4899                2008     manual       110                no       125000   
5808                2008     manual       109                no       100000   
14603               2008     manual       109                no        90000   
21562               2008     manual       109                no       150000   
9720                2008     manual        90                no       150000   
18402               2008     manual       120                no       150000   
18160               2008     manual       120                no       150000   
11163               2008     manual       120                no       150000   
13660               2009     manual        68                no       100000   
10463               2009     manual        68                no       100000   
37008               2009     manual        90                no       150000   
16388               2009     manual        90                no        90000   
6959                2009     manual        73                no        70000   
46964               2010     manual       120                no       150000   
45227               2010     manual        73                no        60000   
41124               2010     manual       111                no        80000   
3039                2010     manual         0                no        80000   
45448               2010     manual        75                no        80000   
28590               2010     manual       156                no        80000   
20298               2010     manual       156                no        70000   
28502               2010     manual        73                no        90000   
13765               2011     manual        90                no       125000   
34958               2011     manual         0                no        70000   
34728               2011     manual       120               yes        40000   
24632               2011     manual       120                no       125000   
26393               2011     manual       163                no       150000   
41176               2011     manual        75                no       150000   
6611                2011     manual        75                no       150000   
48273               2012     manual       110                no        30000   
45299               2012     manual        75                no        20000   
12877               2013     manual       115                no        40000   
10547               2013     manual       156                no        30000   
14984               2013     manual       156                no        30000   
25321               2013     manual        75                no       100000   
45703               2013     manual       200                no        30000   
49716               2013     manual       128                no        80000   
7567                2014     manual       131                no       125000   
47752  